### ISSUE TITLE
🔥Neural network graph API design - meta learning

### ISSUE THREAD

**Issue by lgarithm** [tensorlayer/tensorlayer#394]
In some special case, a graph of the model is important. 
TensorFlow graph is good for visualization but difficult for layer-level processing, so we will provide a graph of layers in TensorLayer fashion.

**Comment by zsdonghao** [comment]
reference:

caffe graph: https://github.com/BVLC/caffe/blob/master/examples/mnist/lenet.prototxt

**Comment by lgarithm** [comment]
![g dot](https://user-images.githubusercontent.com/2044532/38171657-fe3881ca-35d0-11e8-956f-2a6bac105e6d.png)

**Comment by zsdonghao** [comment]
@lgarithm should we use `out_shape` instead of `shape`?

**Comment by lgarithm** [comment]
https://github.com/lgarithm/tensorlayer/commit/51ee5a855d0c83681d3474556e1495226abd97ce changed `shape` to `out_shape`

**Comment by zsdonghao** [comment]
We need all args of the class, otherwise, we can't fully store the behaviour of the layer, then we can't reconstruct the network from the graph.

For example:
- For `DenseLayer`, we need act, n_units.
- For `Conv2d`, we need act, n_filter, filter_size, stride, padding (alternatively, we can get the n_filter, filter_size from the shape of `W_conv2d`, and get the `stride` by comparing the outputs shape the this layer and previous layer.)

ref: https://stackoverflow.com/questions/218616/getting-method-parameter-names-in-python

**Comment by zsdonghao** [comment]
The TL and TF version may need to stored into the graph file as well.

**Comment by fangde** [comment]
Dear all:

that is really exciting discusss.
I've working on this for a while when I developing tensordb.

keras has already implmented this for simle network, they call it  sequential

my advise is maybe we should follow caffe standard, 
overall, the data scheme should be three part

1) the topology
2) node attribute
3) and the attribute them self.


keep in mind we may need to support muli-input and multi-output, and weight sharing.
they are very valulabe for research purpose, which tl targets.


another issue i want to point out is we need tl support  some iterators, for each layer, there is a interface to visit its input and outputs layer.

**Comment by zsdonghao** [comment]
@lgarithm hi, thank you for your great contribution. I just open a PR to make all pooling layer from function to class. This can help all layers to save it's own layer type, which is import for building the network from the graph file.~

**Comment by zsdonghao** [comment]
After having a graph, would we compute the total number of multiplication and add operation in a network? I found it is useful for network design.

**Comment by lgarithm** [comment]
Alternatively, we can compute the **BFLOPs** of each layer, and the whole network, like darknet:
```
layer     filters    size              input                output
    0 conv     32  3 x 3 / 1   416 x 416 x   3   ->   416 x 416 x  32  0.299 BFLOPs
    1 max          2 x 2 / 2   416 x 416 x  32   ->   208 x 208 x  32
    2 conv     64  3 x 3 / 1   208 x 208 x  32   ->   208 x 208 x  64  1.595 BFLOPs
    3 max          2 x 2 / 2   208 x 208 x  64   ->   104 x 104 x  64
    4 conv    128  3 x 3 / 1   104 x 104 x  64   ->   104 x 104 x 128  1.595 BFLOPs
    5 conv     64  1 x 1 / 1   104 x 104 x 128   ->   104 x 104 x  64  0.177 BFLOPs
    6 conv    128  3 x 3 / 1   104 x 104 x  64   ->   104 x 104 x 128  1.595 BFLOPs
    7 max          2 x 2 / 2   104 x 104 x 128   ->    52 x  52 x 128
    8 conv    256  3 x 3 / 1    52 x  52 x 128   ->    52 x  52 x 256  1.595 BFLOPs
    9 conv    128  1 x 1 / 1    52 x  52 x 256   ->    52 x  52 x 128  0.177 BFLOPs
...
```

**Comment by zsdonghao** [comment]
feel free to reopen if you want to discuss.

we just release 2.0.0 version
